lib/fs/folding.go

```diff
@@ -9,41 +9,51 @@ package fs
 import (
 	"strings"
 	"unicode"
-	"unicode/utf8"
 
 	"golang.org/x/text/unicode/norm"
 )
 
+type caseType int
+
+const (
+	asciiLower caseType = iota
+	asciiMixed
+	nonAscii
+)
+
 // UnicodeLowercaseNormalized returns the Unicode lower case variant of s,
 // having also normalized it to normalization form C.
 func UnicodeLowercaseNormalized(s string) string {
-	i := firstCaseChange(s)
-	if i == -1 {
-		return norm.NFC.String(s)
+	switch checkCase(s) {
+	case asciiLower:
+		return s
+	case asciiMixed:
+		return strings.ToLower(s)
+	default:
+		return norm.NFC.String(strings.Map(toLower, s))
 	}
+}
 
-	var rs strings.Builder
-	// WriteRune always reserves utf8.UTFMax bytes for non-ASCII runes,
-	// even if it doesn't need all that space. Overallocate now to prevent
-	// it from ever triggering a reallocation.
-	rs.Grow(utf8.UTFMax - 1 + len(s))
-	rs.WriteString(s[:i])
-
-	for _, r := range s[i:] {
-		rs.WriteRune(unicode.ToLower(unicode.ToUpper(r)))
+func toLower(r rune) rune {
+	if r <= unicode.MaxASCII {
```

Comment:
This could be optimized another ~50% via this bikeshed: https://stackoverflow.com/a/77689444/1432614

Reply:
We're dealing with 4 cases:

* ASCII lower -> no-op
* ASCII mixed -> replace uppercase ones
* Unicode lower -> apply NFC
* Unicode mixed -> replace uppercase ones and apply NFC

The linked logic only offers fast ASCII detection but lacks the upper/lower distinction.